usb: add USB host controller API, support for virtual controller, and experimental host support

jfischer-no opened the pull request:
Add USB host controller API,  MAX3421E driver and shield support, support for virtual controller, and experimental host support.

~~Based/depends on #38679.~~

jfischer-no committed:
drivers: usb: add common layer of UHC API and MAX3421E driver

Add common layer of UHC API and MAX3421E host controller driver.
This implements the bare minimum necessary to communicate with
one peripheral device.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
shields: add Sparkfun MAX3421E shield support

Add Sparkfun MAX3421E shield support.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: add initial USB host support

This is initial support. Necessary to test the UHC driver API,
for the USB support test implementations, and upcoming USBIP support.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
include: drivers: add USB host controller (UHC) driver API header

Add USB host controller (UHC) driver API header.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: usb: add support for virtual USB bus

Add support for virtual USB bus (UVB) intended for use
by virtual UDC and UHC controllers. UVB is not an emulation
of the USB bus or protocol but a simplified service to connect
a virtual host controller to one or more virtual device
controllers.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: uhc: add driver for virtual USB host controller

Add support for virtual USB host controller intended for use
together with virtual bus and virtual device controllers.
This driver is not an emulation of any real host controller.
The driver has initial support for handling control and bulk
transfers.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc: add driver for virtual USB device controller

Add support for virtual USB device controller intended for use
by virtual bus and virtual UHC controllers. This driver is not
an emulation of any real host controller.
The driver has initial support for handling control and bulk
transfers.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: shell: extend new USB support sample by host support

Extend new USB support sample to include experimental host support.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

mateusz- commented on boards/shields/max3421e/doc/index.rst:
```diff
@@ -0,0 +1,58 @@
++-----------------------+-----------------------+-----------------------------+
+| D6                    | not connected         |                             |
++-----------------------+-----------------------+-----------------------------+
+| D7                    | RESET pin             |                             |
++-----------------------+-----------------------+-----------------------------+
+| D8                    | not connected         |                             |
++-----------------------+-----------------------+-----------------------------+
+| D9                    | INTn                  | 24 (Interrupt Output)       |
++-----------------------+-----------------------+-----------------------------+
+| D10                   | SPI CSn               | 34 (SPI Chip Select)        |
++-----------------------+-----------------------+-----------------------------+
+| D11                   | SPI MOSI              | 36 (SPI Serial Data Input)  |
++-----------------------+-----------------------+-----------------------------+
+| D12                   | SPI MISO              | 35 (SPI Serial Data Out)    |
++-----------------------+-----------------------+-----------------------------+
+| D13                   | SPI SCK               | 37 (SPI Serial Clock Input) |
++-----------------------+-----------------------+-----------------------------+
```
Where are the MAX3421E pin numbers from? The [Sparkfun host shield datasheet](https://www.sparkfun.com/datasheets/DevTools/Arduino/USBHostShield-v13.pdf) has different numbers in it for the pins.

jfischer-no replied:
Where do you see differences? These are shield pins not MAX3421E pin numbers.

mateusz- replied:
I see. The third column is labeled "MAX3421E Pin", so I looked at the pins on the MAX.

mateusz- committed:
include: usb_hub: add macros for the USB hub class

Add new usb_hub header file, along with class feature selectors and
request codes.

Signed-off-by: Mateusz Wielgos <mateusz.wielgos@emerson.com>

mateusz- committed:
usb: host: add port power and port reset USB hub features

Add port power and port reset USB hub features.

Signed-off-by: Mateusz Wielgos <mateusz.wielgos@emerson.com>

desowin commented on subsys/usb/host/usbh_ch9.c:
```diff
@@ -0,0 +1,172 @@
+	struct usb_setup_packet req = {
+		.bmRequestType = bmRequestType,
+		.bRequest = bRequest,
+		.wValue = wValue,
+		.wIndex = wIndex,
+		.wLength = wLength,
```
```suggestion
		.wValue = sys_cpu_to_le16(wValue),
		.wIndex = sys_cpu_to_le16(wIndex),
		.wLength = sys_cpu_to_le16(wLength),
```

desowin commented on include/zephyr/usb/usbh.h:
```diff
@@ -0,0 +1,161 @@
+	sys_dlist_t peripherals;
+};
+
+#define USBH_CONROLLER_DEFINE(device_name, uhc_dev)			\
```
nitpick: `USBH_CONROLLER_DEFINE` -> `USBH_CONTROLLER_DEFINE`

desowin commented on subsys/usb/host/usbh_shell.c:
```diff
@@ -0,0 +1,541 @@
+		buf = net_buf_get(&xfer->done, K_NO_WAIT);
+		if (buf) {
+			if (xfer->ep == USB_CONTROL_EP_IN &&
+			    buf->len == sizeof(struct usb_device_descriptor)) {
```
This will treat any descriptor as device descriptor if the size matches. This can be really confusing.

jfischer-no replied:
Yes, will try to improve it until tomorrow.

desowin commented on subsys/usb/host/usbh_shell.c:
```diff
@@ -0,0 +1,541 @@
+const static struct shell *ctx_shell;
+
+static void print_dev_desc(const struct shell *sh,
+			   const struct usb_device_descriptor *const desc)
```
This function is called with byte buffer, not struct itself. Therefore, the appropriate endian swapping (with `sys_le16_to_cpu`) has to be done here before printing.

jfischer-no commented on include/zephyr/drivers/usb/uhc.h:
```diff
@@ -0,0 +1,481 @@
+}
+
+/**
+ * @brief Enable Start of Frame generator
```
uhc_bus_resume() signals resume for at least 20ms, SoF generator should subsequently start within 3ms.
uhc_bus_suspend() actually stops SoF generator and emits an event when it is really suspended.

jfischer-no commented on include/zephyr/drivers/usb/uhc.h:
```diff
@@ -0,0 +1,481 @@
+ * @param[in] xfer   Pointer to UHC transfer
+ *
+ * @return 0 on success, all other values should be treated as error.
+ * @retval -EPERM controller is not initialized
```
IIRC it is inline with UDC API.

jfischer-no commented on include/zephyr/drivers/usb/uhc.h:
```diff
@@ -0,0 +1,481 @@
+ * @param[in] dev    Pointer to device struct of the driver instance
+ *
+ * @return 0 on success, all other values should be treated as error.
+ * @retval -EALREADY controller is not initialized
```
complete uninitialization, as opposite to uhc_init(). This may be a little exaggerated, can be reconsidered later.

jfischer-no commented on drivers/usb/uhc/uhc_common.c:
```diff
@@ -0,0 +1,352 @@
+		goto buf_alloc_error;
+	}
+
+	if (xfer->queued || xfer->claimed) {
```
No this user error should always be prevented.

jfischer-no commented on drivers/usb/uhc/uhc_common.h:
```diff
@@ -0,0 +1,186 @@
+ * @param[in] xfer   Pointer to UHC transfer
+ *
+ * @return 0 on success, all other values should be treated as error.
+ * @retval -EPERM controller is not initialized
```
EPERM is used to be inline with UDC API. I see EHOSTDOWN is very little and only used by networking subsystem.

jfischer-no commented on drivers/usb/uhc/uhc_max3421e.h:
```diff
@@ -0,0 +1,213 @@
+
+/* Register RCVBC */
+#define MAX3421E_REG_RCVBC		6U
+#define MAX3421E_RCVBC_MAX		0x7FU
```
Well bit(7) is always 0, there is no need for bit operations, RCVBC_MAX and SNDBC_MAX are currently not used.

jfischer-no commented on drivers/usb/uhc/uhc_max3421e.c:
```diff
@@ -0,0 +1,1130 @@
+			return ret;
+		}
+
+		net_buf_pull(buf, len);
```
Added comment, I may have to postpone it to the future.

jfischer-no commented on drivers/usb/uhc/uhc_max3421e.c:
```diff
@@ -0,0 +1,1130 @@
+		uint8_t reg[2] = {0, hxfr};
+
+		/* Update DATA PID if transfer parameter changes */
+		max3421e_tgl_update(dev);
```
DATA PID handling is bare minimum to get it working. I very much appreciate your comments, but I would rather not touch it now, or until the implementation for a second real controller is in place.

jfischer-no commented on drivers/usb/uhc/uhc_max3421e.c:
```diff
@@ -0,0 +1,1130 @@
+	int ret = 0;
+
+	if (priv->addr != addr) {
+		priv->addr = addr;
```
I tend towards specific flags, still not sure, added comment.

jfischer-no commented on drivers/usb/uhc/uhc_max3421e.c:
```diff
@@ -0,0 +1,1130 @@
+	if (!xfer->setup) {
+		/* Handle SETUP stage */
+		ret = max3421e_write(dev, MAX3421E_REG_SUDFIFO,
+				     buf->data, MIN(buf->len, 8));
```
>Under no conditions should the USB host stack pass control transfer without 8 bytes setup data (if it does, then the code is horribly broken).

I agree with your objections, but here I would like to keep it that way, for specific purposes. This API will not only be used by USB Host stack.

jfischer-no commented on drivers/usb/uhc/uhc_max3421e.c:
```diff
@@ -0,0 +1,1130 @@
+	return 0;
+}
+
+static const struct uhc_api max3421e_uhc_api = {
```
yesno, I am afraid we should implement only necessary minimum, not sure what it is, and let upper layer manage the rest.

desowin commented on subsys/usb/host/usbh_shell.c:
```diff
@@ -0,0 +1,627 @@
+		buf = net_buf_get(&xfer->done, K_NO_WAIT);
+		if (buf) {
+			if (xfer->ep == USB_CONTROL_EP_IN) {
+				print_desc(ctx_shell, buf);
```
This code assumes that all control requests are reading descriptors which is not true for `FOOBAZ_VREQ_IN`. To solve this, the request context should be somehow preserved. Just a FIXME comment for now seems sufficient though.

jfischer-no replied:
Added comment

desowin commented on subsys/usb/host/usbh_shell.c:
```diff
@@ -0,0 +1,627 @@
+
+static void print_desc(const struct shell *sh, const struct net_buf *const buf)
+{
+	struct usb_desc_header *head = (void *)buf->data;
```
This function doesn't sanitize received data length. It shouldn't really cause any issues other than simply printing uninitialized data if device responds with outright incorrect descriptor (too short to be descriptor of given type).

jfischer-no replied:
Added net_buf length checks.

desowin reviewed (approved):
Single device only with appropriate FIXME/TODO comments related to the issues that will inevitably show up when implementing hub support.
